### DISCUSSION TITLE
Order of change application

### DISCUSSION THREAD

**Discussion by spothound** [fluxcd/flux2#2058] [Q&A] [answered]
Hello there, 

I have an intuition on how does FluxV2 reconciliation process work but I cannot find anything in the documentation that confirms this intuition. I have this doubt: What happens if I define two resources so one required the other to exist? For example, I define a new namespace in a YAML config file and then a pod that will be deployed in this namespace in another different YAML file. How does FluxCD know which one should be created first?

I guess that what if does is trying to apply both required changes in an "arbitrary order" and, if one of them fails, ignore the error and try to apply the change again after applying the rest. 

So, in this case, Flux may try to create the pod before creating the namespace, this would cause an error that would be ignored and then, Flux would create the namespace and try to create the pod again with success. 

Is that correct? If not, how does Flux know when there is a dependency between resources? Can I specify the order?

If you can point me to any documentation talking about this I would appreciate it.

Best regards.

**Comment by yebyen** [comment, marked as answer]
Hello!

The apply function in kustomize-controller:
https://github.com/fluxcd/kustomize-controller/blob/5b820ec1feba5251d795128cd96061273f0f03ca/controllers/kustomization_controller.go#L656-L657

is divided into Stage One ☝️ and Stage Two 👇 

https://github.com/fluxcd/kustomize-controller/blob/5b820ec1feba5251d795128cd96061273f0f03ca/controllers/kustomization_controller.go#L659-L660

So, if the issue is that you need CRDs applied before Custom Resources that they represent, or similarly that you need a Namespace applied before the namespaced resources that go inside, Flux's Kustomize Controller is already smart enough to handle that without any help.

If there are more complex dependencies then they may need to be represented in more than one Flux Kustomization. You can control the set of resources that a Kustomization will apply by pointing it at a particular folder in the GitRepository source, with`spec.path`, and you can enforce an order that Kustomization resources are applied and must become ready in by using `spec.dependsOn` in the Flux Kustomization resource, and additionally consider adding `spec.wait` to ensure that resources are not considered ready before they are actually live and Ready according to Kstatus.

I would also mention  `spec.timeout`, which should be set if you want to see an error in a failing Kustomization's status reports in lesser time than `spec.interval` (which can be important if you set `spec.interval` to a high number), and also `spec.retryInterval` which can ensure that failed states are retried more often than `spec.interval`.

When there are more than a handful of Kustomization resources, all of these values can make sense to consider separately.

**Comment by stefanprodan** [comment]
> I guess that what if does is trying to apply both required changes in an "arbitrary order" and, if one of them fails, ignore the error and try to apply the change again after applying the rest.

The order is not arbitrary, we use a ordering algorithm based on kind and resource name. Flux doesn’t apply a commit if any of the resources are invalid, we do a server-side apply dry-run to validate all resources before applying them.

**Reply by spothound** [reply]
Thank you both guys! now I have a better understanding of how Fluxcd works, I also understood today what @stefanprodan said about flux not playing a commit it has any invalid resource... I was struggling to get changes applied by flux because I had errors in previously committed YAML files and I thought they couldn't interfere in others' file changes. Now I have tried to fix those errors and my new changes are being applied. 

Also, I wanted to know how the reconciliation was implemented and I didn't know about those spec options. It seems that I still have a lot of things to learn from Flux!

Again, thank you guys!

**Reply by stefanprodan** [reply]
@spothound if you want find out more about how reconciliation works and which options are available, please see the API spec: https://fluxcd.io/docs/components/kustomize/kustomization/
